### ISSUE TITLE
Microsoft.Azure.Cosmos.CosmosException

### ISSUE THREAD

**Issue by fecalemos** [Azure/azure-cosmos-dotnet-v3#945]
When I am trying to run a console and trying to create using,
Database database = await client.CreateDatabaseIfNotExistsAsync(databaseId);
in InitializeCosmosClientInstanceAsync, I am getting the below exception

> ServiceUnavailable error occured: Response status code does not indicate success: 503 Substatus: 0 Reason: (The request failed because the client was unable to establish connections to 3 endpoints across 1 regions. Please check for client resource starvation issues and verify connectivity between client and server.

**Comment by j82w** [comment]
1. Are you using the local emulator or a real cosmos db account?
2. Did you verify the account name and password are correct?

**Comment by fecalemos** [comment]
1. I am using real cosmos db account.
2. Yes, I have azure portal account.

**Comment by j82w** [comment]
Did you verify that the Cosmos DB account name and key are correct when passing it into the CosmosClient?  The other possibility is a firewall is blocking the [required ports](https://docs.microsoft.com/en-us/azure/cosmos-db/performance-tips#networking).

<img width="561" alt="accountInfo" src="https://user-images.githubusercontent.com/8868107/67855636-070dd780-fad0-11e9-8cbc-3cd71efa6ed4.PNG">

**Comment by fecalemos** [comment]
Yes, I verify that the CosmosDB account name and key are correct. 
the new cosmos db instance is being created below:

![image](https://user-images.githubusercontent.com/4366391/67971465-93f98380-fbeb-11e9-8ddf-d6883c1b0b1c.png)


error output result:
> "Response status code does not indicate success: 503 Substatus: 0 Reason: (The request failed because the client was unable to establish connections to 3 endpoints across 1 regions. Please check for client resource starvation issues and verify connectivity between client and server.ActivityId: eac02104-2513-4f88-b5cd-64cae83e87e5, {\"responseStatisticsList\":[{\"RequestResponseTime\":\"2019-10-31T17:28:33.5877968Z\",\"StoreResult\":{\"LSN\":-1,\"QuorumAckedLSN\":-1,\"GlobalCommittedLSN\":-1,\"NumberOfReadRegions\":-1,\"ItemLSN\":-1,\"UsingLocalLSN\":false,\"RequestCharge\":0.0,\"CurrentReplicaSetSize\":-1,\"CurrentWriteQuorum\":-1,\"IsValid\":false,\"IsGoneException\":true,\"IsNotFoundException\":false,\"IsInvalidPartitionException\":false,\"StorePhysicalAddress\":\"rntbd://cdb-ms-prod-brazilsouth1-fd3.documents.azure.com:11002/apps/0418e967-b1b2-444d-a6e7-4223c524208e/services/1398baeb-2ca2-42b6-847f-e4ea04f5ec3f/partitions/57448983-4eec-4dc6-9d4e-75fff932854a/replicas/132169546034711664s/\",\"IsClientCpuOverloaded\":false},\"RequestResourceType\":0,\"RequestOperationType\":2},{\"RequestResponseTime\":\"2019-10-31T17:28:33.5884183Z\",\"StoreResult\":{\"LSN\":-1,\"QuorumAckedLSN\":-1,\"GlobalCommittedLSN\":-1,\"NumberOfReadRegions\":-1,\"ItemLSN\":-1,\"UsingLocalLSN\":false,\"RequestCharge\":0.0,\"CurrentReplicaSetSize\":-1,\"CurrentWriteQuorum\":-1,\"IsValid\":false,\"IsGoneException\":true,\"IsNotFoundException\":false,\"IsInvalidPartitionException\":false,\"StorePhysicalAddress\":\"rntbd://cdb-ms-prod-brazilsouth1-fd3.documents.azure.com:11006/apps/0418e967-b1b2-444d-a6e7-4223c524208e/services/1398baeb-2ca2-42b6-847f-e4ea04f5ec3f/partitions/57448983-4eec-4dc6-9d4e-75fff932854a/replicas/132169833988797452s/\",\"IsClientCpuOverloaded\":false},\"RequestResourceType\":0,\"RequestOperationType\":2},{\"RequestResponseTime\":\"2019-10-31T17:28:39.5281249Z\",\"StoreResult\":{\"LSN\":-1,\"QuorumAckedLSN\":-1,\"GlobalCommittedLSN\":-1,\"NumberOfReadRegions\":-1,\"ItemLSN\":-1,\"UsingLocalLSN\":false,\"RequestCharge\":0.0,\"CurrentReplicaSetSize\":-1,\"CurrentWriteQuorum\":-1,\"IsValid\":false,\"IsGoneException\":true,\"IsNotFoundException\":false,\"IsInvalidPartitionException\":false,\"StorePhysicalAddress\":\"rntbd://cdb-ms-prod-brazilsouth1-fd3.documents.azure.com:11309/apps/0418e967-b1b2-444d-a6e7-4223c524208e/services/1398baeb-2ca2-42b6-847f-e4ea04f5ec3f/partitions/57448983-4eec-4dc6-9d4e-75fff932854a/replicas/132169952292095127s/\",\"IsClientCpuOverloaded\":false},\"RequestResourceType\":0,\"RequestOperationType\":2},{\"RequestResponseTime\":\"2019-10-31T17:28:45.5754875Z\",\"StoreResult\":{\"LSN\":-1,\"QuorumAckedLSN\":-1,\"GlobalCommittedLSN\":-1,\"NumberOfReadRegions\":-1,\"ItemLSN\":-1,\"UsingLocalLSN\":false,\"RequestCharge\":0.0,\"CurrentReplicaSetSize\":-1,\"CurrentWriteQuorum\":-1,\"IsValid\":false,\"IsGoneException\":true,\"IsNotFoundException\":false,\"IsInvalidPartitionException\":false,\"StorePhysicalAddress\":\"rntbd://cdb-ms-prod-brazilsouth1-fd3.documents.azure.com:11006/apps/0418e967-b1b2-444d-a6e7-4223c524208e/services/1398baeb-2ca2-42b6-847f-e4ea04f5ec3f/partitions/57448983-4eec-4dc6-9d4e-75fff932854a/replicas/132169833988797452s/\",\"IsClientCpuOverloaded\":false},\"RequestResourceType\":0,\"RequestOperationType\":2},{\"RequestResponseTime\":\"2019-10-31T17:28:45.5756445Z\",\"StoreResult\":{\"LSN\":-1,\"QuorumAckedLSN\":-1,\"GlobalCommittedLSN\":-1,\"NumberOfReadRegions\":-1,\"ItemLSN\":-1,\"UsingLocalLSN\":false,\"RequestCharge\":0.0,\"CurrentReplicaSetSize\":-1,\"CurrentWriteQuorum\":-1,\"IsValid\":false,\"IsGoneException\":true,\"IsNotFoundException\":false,\"IsInvalidPartitionException\":false,\"StorePhysicalAddress\":\"rntbd://cdb-ms-prod-brazilsouth1-fd3.documents.azure.com:11309/apps/0418e967-b1b2-444d-a6e7-4223c524208e/services/1398baeb-2ca2-42b6-847f-e4ea04f5ec3f/partitions/57448983-4eec-4dc6-9d4e-75fff932854a/replicas/132169952292095127s/\",\"IsClientCpuOverloaded\":false},\"RequestResourceType\":0,\"RequestOperationType\":2},{\"RequestResponseTime\":\"2019-10-31T17:28:51.5623584Z\",\"StoreResult\":{\"LSN\":-1,\"QuorumAckedLSN\":-1,\"GlobalCommittedLSN\":-1,\"NumberOfReadRegions\":-1,\"ItemLSN\":-1,\"UsingLocalLSN\":false,\"RequestCharge\":0.0,\"CurrentReplicaSetSize\":-1,\"CurrentWriteQuorum\":-1,\"IsValid\":false,\"IsGoneException\":true,\"IsNotFoundException\":false,\"IsInvalidPartitionException\":false,\"StorePhysicalAddress\":\"rntbd://cdb-ms-prod-brazilsouth1-fd3.documents.azure.com:11002/apps/0418e967-b1b2-444d-a6e7-4223c524208e/services/1398baeb-2ca2-42b6-847f-e4ea04f5ec3f/partitions/57448983-4eec-4dc6-9d4e-75fff932854a/replicas/132169546034711664s/\",\"IsClientCpuOverloaded\":false},\"RequestResourceType\":0,\"RequestOperationType\":2},{\"RequestResponseTime\":\"2019-10-31T17:28:57.5379847Z\",\"StoreResult\":{\"LSN\":-1,\"QuorumAckedLSN\":-1,\"GlobalCommittedLSN\":-1,\"NumberOfReadRegions\":-1,\"ItemLSN\":-1,\"UsingLocalLSN\":true,\"RequestCharge\":0.0,\"CurrentReplicaSetSize\":-1,\"CurrentWriteQuorum\":-1,\"IsValid\":false,\"IsGoneException\":true,\"IsNotFoundException\":false,\"IsInvalidPartitionException\":false,\"StorePhysicalAddress\":\"rntbd://cdb-ms-prod-brazilsouth1-fd3.documents.azure.com:11304/apps/0418e967-b1b2-444d-a6e7-4223c524208e/services/1398baeb-2ca2-42b6-847f-e4ea04f5ec3f/partitions/57448983-4eec-4dc6-9d4e-75fff932854a/replicas/132145401301671179p/\",\"IsClientCpuOverloaded\":false},\"RequestResourceType\":0,\"RequestOperationType\":2},{\"RequestResponseTime\":\"2019-10-31T17:28:57.6443114Z\",\"StoreResult\":{\"LSN\":-1,\"QuorumAckedLSN\":-1,\"GlobalCommittedLSN\":-1,\"NumberOfReadRegions\":-1,\"ItemLSN\":-1,\"UsingLocalLSN\":true,\"RequestCharge\":0.0,\"CurrentReplicaSetSize\":-1,\"CurrentWriteQuorum\":-1,\"IsValid\":false,\"IsGoneException\":true,\"IsNotFoundException\":false,\"IsInvalidPartitionException\":false,\"StorePhysicalAddress\":\"rntbd://cdb-ms-prod-brazilsouth1-fd3.documents.azure.com:11304/apps/0418e967-b1b2-444d-a6e7-4223c524208e/services/1398baeb-2ca2-42b6-847f-e4ea04f5ec3f/partitions/57448983-4eec-4dc6-9d4e-75fff932854a/replicas/132145401301671179p/\",\"IsClientCpuOverloaded\":false},\"RequestResourceType\":0,\"RequestOperationType\":2},{\"RequestResponseTime\":\"2019-10-31T17:29:03.5702927Z\",\"StoreResult\":{\"LSN\":-1,\"QuorumAckedLSN\":-1,\"GlobalCommittedLSN\":-1,\"NumberOfReadRegions\":-1,\"ItemLSN\":-1,\"UsingLocalLSN\":false,\"RequestCharge\":0.0,\"CurrentReplicaSetSize\":-1,\"CurrentWriteQuorum\":-1,\"IsValid\":false,\"IsGoneException\":true,\"IsNotFoundException\":false,\"IsInvalidPartitionException\":false,\"StorePhysicalAddress\":\"rntbd://cdb-ms-prod-brazilsouth1-fd3.documents.azure.com:11006/apps/0418e967-b1b2-444d-a6e7-4223c524208e/services/1398baeb-2ca2-42b6-847f-e4ea04f5ec3f/partitions/57448983-4eec-4dc6-9d4e-75fff932854a/replicas/132169833988797452s/\",\"IsClientCpuOverloaded\":false},\"RequestResourceType\":0,\"RequestOperationType\":2},{\"RequestResponseTime\":\"2019-10-31T17:29:03.5704459Z\",\"StoreResult\":{\"LSN\":-1,\"QuorumAckedLSN\":-1,\"GlobalCommittedLSN\":-1,\"NumberOfReadRegions\":-1,\"ItemLSN\":-1,\"UsingLocalLSN\":false,\"RequestCharge\":0.0,\"CurrentReplicaSetSize\":-1,\"CurrentWriteQuorum\":-1,\"IsValid\":false,\"IsGoneException\":true,\"IsNotFoundException\":false,\"IsInvalidPartitionException\":false,\"StorePhysicalAddress\":\"rntbd://cdb-ms-prod-brazilsouth1-fd3.documents.azure.com:11309/apps/0418e967-b1b2-444d-a6e7-4223c524208e/services/1398baeb-2ca2-42b6-847f-e4ea04f5ec3f/partitions/57448983-4eec-4dc6-9d4e-75fff932854a/replicas/132169952292095127s/\",\"IsClientCpuOverloaded\":false},\"RequestResourceType\":0,\"RequestOperationType\":2}],\"supplementalResponseStatisticsList\":[],\"addressResolutionStatistics\":{\"dd47ab63-cc5f-41b8-b950-e692bfa8bc03\":{\"StartTime\":\"2019-10-31T17:28:27.6957657Z\",\"EndTime\":\"2019-10-31T17:28:27.8457547Z\",\"TargetEndpoint\":\"https://glosignh001.documents.azure.com//addresses/?$resolveFor=dbs&$filter=protocol eq rntbd\"},\"2184672e-20db-432c-98a6-cc1fd578b225\":{\"StartTime\":\"2019-10-31T17:28:39.5378222Z\",\"EndTime\":\"2019-10-31T17:28:39.6526205Z\",\"TargetEndpoint\":\"https://glosignh001.documents.azure.com//addresses/?$resolveFor=dbs&$filter=protocol eq rntbd\"},\"4d75f534-1acd-469a-9657-12bf8b056108\":{\"StartTime\":\"2019-10-31T17:28:57.5435073Z\",\"EndTime\":\"2019-10-31T17:28:57.6353352Z\",\"TargetEndpoint\":\"https://glosignh001.documents.azure.com//addresses/?$resolveFor=dbs&$filter=protocol eq rntbd\"},\"dfce87a6-0aa1-4c93-8d28-9e0c14116596\":{\"StartTime\":\"2019-10-31T17:28:57.6698287Z\",\"EndTime\":\"2019-10-31T17:28:57.7675621Z\",\"TargetEndpoint\":\"https://glosignh001.documents.azure.com//addresses/?$resolveFor=dbs&$filter=protocol eq rntbd\"}},\"ContactedReplicas\":[],\"FailedReplicas\":[\"rntbd://cdb-ms-prod-brazilsouth1-fd3.documents.azure.com:11002/apps/0418e967-b1b2-444d-a6e7-4223c524208e/services/1398baeb-2ca2-42b6-847f-e4ea04f5ec3f/partitions/57448983-4eec-4dc6-9d4e-75fff932854a/replicas/132169546034711664s/\",\"rntbd://cdb-ms-prod-brazilsouth1-fd3.documents.azure.com:11006/apps/0418e967-b1b2-444d-a6e7-4223c524208e/services/1398baeb-2ca2-42b6-847f-e4ea04f5ec3f/partitions/57448983-4eec-4dc6-9d4e-75fff932854a/replicas/132169833988797452s/\",\"rntbd://cdb-ms-prod-brazilsouth1-fd3.documents.azure.com:11309/apps/0418e967-b1b2-444d-a6e7-4223c524208e/services/1398baeb-2ca2-42b6-847f-e4ea04f5ec3f/partitions/57448983-4eec-4dc6-9d4e-75fff932854a/replicas/132169952292095127s/\"],\"RegionsContacted\":[\"https://glosignh001.documents.azure.com:443/\"],\"RequestLatency\":\"00:00:36.0425465\",\"IsCpuOverloaded\":false}, Windows/10.0.18362 cosmos-netstandard-sdk/3.3.0)."

I am seeing firewall on my azure portal that allows access from **all networks**

**Comment by j82w** [comment]
Can you try using Gateway mode in the CosmosClientOptions?

**Comment by fecalemos** [comment]
@j82w you are right because that is gateway mode but I tested two different ways:
1. .Net Core with console app are OK, when CosmosDB read the item with id and partitionKey successfully.
2. .Net Core with web api are NOT OK, can't postman access the endpoint? I got the error message (_Object reference not set to an instance of an object_) when CosmosDB tries to read the item using ReadItemAsync with id and partitionKey. See the print below.
![print_azure_exception_code01](https://user-images.githubusercontent.com/4366391/68310468-d3621d00-008e-11ea-9fbf-646f9f8d19fb.png)
